crates/sui-core/src/authority_active/gossip/node_sync.rs

```diff
@@ -0,0 +1,435 @@
+        let digest = digests.transaction;
+        if let Some(c) = self.node_sync_store.get_cert_and_effects(&digest)? {
+            return Ok(c);
+        }
+
+        let (tx, mut rx) = self.pending_downloads.wait(&digest).await;
+        // Only start the download if there are no other concurrent downloads.
+        if let Some(tx) = tx {
+            let aggregator = self.aggregator.clone();
+            let digests = *digests;
+            let peer = *peer;
+            let node_sync_store = self.node_sync_store.clone();
+            tokio::task::spawn(async move {
+                if let Err(error) =
+                    tx.send(Self::download_impl(peer, aggregator, &digests, node_sync_store).await)
+                {
+                    error!(?digest, ?peer, ?error, "Could not broadcast cert response");
+                }
+            });
+        }
+
+        let _ = rx
+            .recv()
+            .await
+            .map_err(|e| SuiError::GenericAuthorityError {
+                error: format!("{:?}", e),
+            })??;
+
+        self.node_sync_store
+            .get_cert_and_effects(&digest)?
+            .ok_or_else(|| SuiError::GenericAuthorityError {
+                error: format!(
+                    "cert/effects for {:?} should have been in the node_sync_store",
+                    digest
+                ),
+            })
+    }
+
+    async fn download_impl(
+        peer: AuthorityName,
+        aggregator: Arc<AuthorityAggregator<A>>,
+        digests: &ExecutionDigests,
+        node_sync_store: Arc<NodeSyncStore>,
+    ) -> SuiResult {
+        let digest = digests.transaction;
+
+        // TODO: Add a function to AuthorityAggregator to try multiple validators - even
+        // though we are fetching the cert/effects from the same validator that sent us the tx
+        // digest, and even though we know the cert is final, any given validator may be byzantine
+        // and refuse to give us the cert and effects.
+        let client = aggregator.clone_client(&peer);
+        let resp = client
+            .handle_transaction_and_effects_info_request(digests)
+            .await
+            .expect("TODO: need to use authority aggregator to download cert");
+
+        let cert = resp.certified_transaction.ok_or_else(|| {
+            info!(?digest, ?peer, "validator did not return cert");
+            SuiError::GenericAuthorityError {
+                error: format!("validator did not return cert for {:?}", digest),
+            }
+        })?;
+
+        let effects = resp.signed_effects.ok_or_else(|| {
+            info!(?digest, ?peer, "validator did not return effects");
+            SuiError::ByzantineAuthoritySuspicion { authority: peer }
+        })?;
+
+        node_sync_store.store_cert_and_effects(&digest, &(cert, effects))?;
+
+        Ok(())
+    }
+}
+
+#[derive(Clone)]
+pub struct NodeSyncDigestHandler {
+    _sync_join_handle: Arc<JoinHandle<()>>,
```

Comment:
you technically don't need to hold on to this. Since the task is meant to be long running, the way you "stop" it would be to just drop the sender. Then the channel closes and the other task will clean itself up.

Reply:
It seems to be a common pattern to keep these join handles around, but I can remove it if you like.